Add support for TLS v1.3

windowsair opened the pull request:
TLS v1.3 is available on windows 10.0.17763 and above version. #83 

The associated unittest needs further modification, I'll upload it later.

windowsair commented:
Okay, now it is ready to review :)

laudrup commented on include/wintls/detail/sspi_handshake.hpp:
```diff
@@ -19,13 +19,61 @@
+    DWORD               cTlsParameters;
+    PTLS_PARAMETERS     pTlsParameters;
+} SCH_CREDENTIALS, * PSCH_CREDENTIALS;
+
```
Why are all these C types needed here? Aren't they already defined in some of the SSPI headers already included.

If they are really needed they definitely shouldn't be here.

windowsair replied:
> Minimum supported client	Windows 10 1809 [desktop apps only]

Doing so helps us to compile on lower versions of Windows.

laudrup replied:
> > Minimum supported client	Windows 10 1809 [desktop apps only]
> 
> Doing so helps us to compile on lower versions of Windows.

OK. Then please move it to it's own header and only include if compiling on a version of Windows where it's required. This should probably be in `sspi_types.h`.

laudrup commented on include/wintls/detail/sspi_handshake.hpp:
```diff
@@ -74,25 +115,67 @@ class sspi_handshake {
-      creds.cCreds = 1;
-      creds.paCred = &server_cert;
-    }
+    bool is_tlsv13 = !!(static_cast<int>(context_.method_) & static_cast<int>(method::tlsv13));
```
I'm not really a fan of this C type code as it depends on the underlying value of the enum and is a bit hard to read. Please consider using an inline lambda similar to what is being done for the `usage` variable above.

laudrup commented on include/wintls/detail/sspi_handshake.hpp:
```diff
@@ -51,17 +99,10 @@ class sspi_handshake {
-    if (check_revocation_) {
-      creds.dwFlags |= SCH_CRED_REVOCATION_CHECK_CHAIN_EXCLUDE_ROOT;
-    }
+    void *pCred;
```
Please avoid hungarian notation and please initialize variables.

laudrup commented on include/wintls/detail/sspi_handshake.hpp:
```diff
@@ -74,25 +115,67 @@ class sspi_handshake {
+        credentials.cCreds = 1;
+        credentials.paCred = &server_cert;
+      }
     }
```
Most of the code is duplicated in the different if/else branches here (`is_tlsv3`).

Please find a way to refactor this to avoid that duplication.

laudrup commented on test/tls_record.cpp:
```diff
@@ -77,8 +84,78 @@ tls_handshake::message_type read_message(tls_handshake::handshake_type t, net::c
+      return tls_extension::common{buffer, size};
+    default:
+      return tls_extension::common{buffer, size};
+  }
```
I might be reading this the wrong way, but it definitely looks like this could more or less be reduced to:

`return tls_extension::common{buffer, size};`

Also, if you are going to switch on all the enum values, then please leave out the default case. That way the compiler can warn if an enum value isn't handled which is definitely nice.

So maybe this could be written as:

```
 switch (t) {
    case tls_extension::extension_type::server_name:
    case tls_extension::extension_type::max_fragment_length:
    case tls_extension::extension_type::status_request:
    // Remaing enum values follow...
     return tls_extension::common{buffer, size};
 }
WINTLS_UNREACHABLE_RETURN(0);
```

laudrup reviewed (changes requested):
Thanks a lot for this very useful improvement.

Very good job an extending the unit tests so they also test that this actually does handle TLS v1.3 (I assume it does so :-)).

Please address my comments though.

If you don't have the time or don't want to make these changes yourself then I'll be happy to do so myself when time permits. Your contribution is much appreciated no matter what.

windowsair commented on include/wintls/detail/sspi_handshake.hpp:
```diff
@@ -19,13 +19,61 @@
 
 #include <wintls/handshake_type.hpp>
 
+#include <SubAuth.h>
```
struct member `UNICODE_STRING` in `CRYPTO_SETTINGS` need it.

> To use the SCH_CREDENTIALS structure, define SCHANNEL_USE_BLACKLISTS along with UNICODE_STRING and PUNICODE_STRING. Alternatively, include Ntdef.h, SubAuth.h or Winternl.h.

laudrup replied:
OK. Then please move it so it's included with the other SSPI related headers.

windowsair commented:
There are conflicting codes. But the code in mainline doesn't seem to pass CI. Do I need to wait for the mainline code to be fixed first?

laudrup commented:
> There are conflicting codes. But the code in mainline doesn't seem to pass CI. Do I need to wait for the mainline code to be fixed first?

As far as I can tell the errors were related to a problem with the Windows images used by all Github CI actions, not just this repository, so you should be able to fix the conflicts and rebase without any issues now.

windowsair commented:
> > There are conflicting codes. But the code in mainline doesn't seem to pass CI. Do I need to wait for the mainline code to be fixed first?
> 
> As far as I can tell the errors were related to a problem with the Windows images used by all Github CI actions, not just this repository, so you should be able to fix the conflicts and rebase without any issues now.

Unit test works fine. But coverage and coverity task failed :(

laudrup commented:
> Unit test works fine. But coverage and coverity task failed :(

The coverage and coverity jobs require access tokens which are only available on my github repository. I assume you could generate your own accounts on codecov and coverity if you really wanted to, but that'd hardly be worth it.

Just ignore that for now and please focus on addressing the issues I've pointed out instead.

The coverage and coverity jobs are only being run on the master branch anyway so they won't stop this PR from being merged.

windowsair commented:
> > Unit test works fine. But coverage and coverity task failed :(
> 
> The coverage and coverity jobs require access tokens which are only available on my github repository. I assume you could generate your own accounts on codecov and coverity if you really wanted to, but that'd hardly be worth it.
> 
> Just ignore that for now and please focus on addressing the issues I've pointed out instead.
> 
> The coverage and coverity jobs are only being run on the master branch anyway so they won't stop this PR from being merged.

Okay, it;s ready for review

laudrup commented on include/wintls/detail/sspi_types.hpp:
```diff
@@ -29,4 +29,51 @@
 #define UNICODE
 #endif // WINTLS_UNICODE_UNDEFINED
 
+#include <SubAuth.h>
```
Please move this up between the `#ifdefs` where the two other includes are. Hopefully that doesn't cause any issues. Otherwise we should look into that.

laudrup commented on include/wintls/detail/sspi_types.hpp:
```diff
@@ -29,4 +29,51 @@
+    DWORD               cTlsParameters;
+    PTLS_PARAMETERS     pTlsParameters;
+} SCH_CREDENTIALS, * PSCH_CREDENTIALS;
+
```
You mentioned that these structs only needed to be defined when building on some older versions of Windows.

Could you please move them to an appropriately named header file and then only include that file if we are building on a version of Windows where they are needed?

windowsair replied:
Already moved to `sspi_compat.hpp`

laudrup reviewed (changes requested):
Sorry for being late with this review.

It looks very good. Please address my last comments and then I think this could be merged.

I also made some changes to the master branch so the coverity and coverage jobs should no runner long on forks of this repository. You made me aware of that issue so thanks a lot for that as well. You might want to rebase your fork on this repos master branch.

laudrup commented:
Looks great.

Only one minor comment left then I think should be ready to merge and I should probably also make a new release with this.

Thanks a lot.

laudrup commented:
@windowsair 

Seems like I forgot to actually submit my final comment. After that I think we should merge this.

Thanks.

laudrup commented:
Great and thanks a lot.

Sorry for the slow response time.
